### ISSUE TITLE
Configuration file / settings not persisted with IntelliJ IDEA 2020.1

### ISSUE THREAD

**Issue by C-Otto** [jshiell/checkstyle-idea#476]
When I restart IntelliJ IDEA I only have the two default configurations. My own configuration (using Checkstyle 8.1) is lost.

IntelliJ IDEA 2020.1 EAP (Ultimate Edition)
Build #IU-201.5616.10, built on February 19, 2020
IntelliJ IDEA EAP User
Expiration date: March 20, 2020
Runtime version: 11-ea+105 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 5.4.0-0.bpo.3-amd64
GC: ParNew, ConcurrentMarkSweep
Memory: 3891M
Cores: 16
Registry: 
Non-Bundled Plugins: com.andrey4623.rainbowcsv, com.github.b3er.idea.plugins.arc.browser, CheckStyle-IDEA, Error-prone plugin, Lombook Plugin, com.jetbrains.ChooseRuntime, io.protostuff.protostuff-jetbrains-plugin, net.seesharpsoft.intellij.plugins.csv, Pythonid

Plugin version 5.35.6

**Comment by C-Otto** [comment]
Checkstyle also tells me it was updated on every IntelliJ IDEA reboot.

**Comment by jshiell** [comment]
Thanks for the report - the update notification relies on app-scoped preferences to track if it has displayed, so that's consistent.

The bad news is that config persistence is automagic by the IDE, so any changes may require upgrading the SDK and breaking backwards compatibility. Or it could just be an EAP bug. Either way, I fear this probably won't get looked at until 2020.1 is in RC at the least.

**Comment by C-Otto** [comment]
https://youtrack.jetbrains.com/issue/IDEA-235657

**Comment by C-Otto** [comment]
I'm very grateful for the plugin and I appreciate the effort. Please allow me to ask for more, though. The EAP for 2020.1 has been released end of January. As such, I've been unable to use the plugin for around two months. Other developers might have the same issue.

Furthermore, as soon as 2020.1 is released or in RC, it still takes some time to investigate the issue and update the code.

If you're eager to fix this at any time, please think about fixing it as early as possible. In my experience the EAPs don't introduce back-and-forth changes often, especially for plugin related code (which, of course, is depended on by many plugins.)

**Comment by jshiell** [comment]
Your request is entirely fair - the problem is more than such changes often break backwards compatibility, and hence force me to drop support for older versions. My past experience is that this tends to affect more people than are using the EAP. Even if it wasn't for this particular issue, there are certainly deprecations in the 2020 SDK (and even the 2019) SDK that will require me to drop 2018 support, so again it's just trying to find the balance of when to switch without detrimentally affecting too many users.

But you do make a very valid case, and the 2020 release has just gone into beta from memory, so it's definitely increasing in priority. I'll try and have a look in the not too distant, COVID permitting 😄

**Comment by jshiell** [comment]
Hmm. Initial testing with 201.6487.11 confirms that reading the preferences seems broken. Upon opening an existing project the state is reset to the default; and this is written to disk, overwriting old settings. Modifying the settings does appear to work, and is persisted to disk correctly. However, upon closing and opening the project again the settings are again reset to default, despite the file on disk being current.

TL;DR writing preferences seems to be fine, however no read of the file seems to be happing.

**Comment by jshiell** [comment]
This is, to use a technical term, "fucking weird". According to [the docs](https://www.jetbrains.org/intellij/sdk/docs/basics/persisting_state_of_components.html#implementing-the-state-class) bean properties should be persisted. And indeed they are in older versions. However, in 2020.1 the property seems to persist, but is never restored.

The solution is cargo cult - switch to using an annotated private variable, which should work exactly the same. However, for whatever reason, it does work in 2020.1 in both directions. And best of all, it appears to be valid with the 2018 SDK, so we can continue without breaking backwards compatibility.

So I suspect it is an IDEA bug; however there's a workaround.

Could you please [try this test build](https://www.dropbox.com/s/jqgbofrikdldfpn/CheckStyle-IDEA-5.35.8-pre.zip?dl=0) and verify if this fixes it for you? Note that the app-level persistence isn't fixed yet, so you'll still get nagged about being updated on every start, but it should be enough to validate the fix (and fix the important part). If it does work (and I expect it to, based on my testing) then I'll fix the app-level persistence and get it released ASAP. Thank you!

**Comment by C-Otto** [comment]
Can confirm, it works! Thank you :)

**Comment by jshiell** [comment]
Smashing - just released. Thanks for your help!
